meilisearch-core/src/facets.rs

```diff
@@ -0,0 +1,342 @@
+                                return Err(FacetError::EmptyArray);
+                            }
+                            let mut inner = Vec::new();
+                            for expr in or_exprs {
+                                match expr {
+                                    Value::String(s) => inner.push(FacetKey::from_str(
+                                        &s,
+                                        schema,
+                                        attributes_for_faceting,
+                                    )?),
+                                    bad_value => {
+                                        return Err(FacetError::UnexpectedToken { expected: &["String"], found: bad_value.to_string() })
+                                    }
+                                }
+                            }
+                            filter.push(Either::Left(inner));
+                        }
+                        bad_value => return Err(FacetError::UnexpectedToken { expected: &["String", "Array"], found: bad_value.to_string() }),
+                    }
+                }
+                return Ok(Self(filter));
+            }
+            bad_value => Err(FacetError::UnexpectedToken { expected: &["Array"], found: bad_value.to_string() }),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+#[repr(C)]
+pub struct FacetKey(FieldId, String);
+
+impl FacetKey {
+    pub fn new(field_id: FieldId, value: String) -> Self {
+        let value = match value.cow_to_lowercase() {
+                Cow::Borrowed(_) => value,
+                Cow::Owned(s) => s,
+        };
+        Self(field_id, value)
+    }
+
+    fn from_str(
+        s: &str,
+        schema: &Schema,
+        attributes_for_faceting: &[FieldId],
+    ) -> Result<Self, FacetError> {
+        let mut split = s.splitn(2, ":");
+        let key = split
+            .next()
+            .ok_or_else(|| FacetError::InvalidFormat(s.to_string()))?;
+        let field_id = schema
+            .id(key)
+            .ok_or_else(|| FacetError::AttributeNotFound(key.to_string()))?;
+
+        if !attributes_for_faceting.contains(&field_id) {
+            return Err(FacetError::AttributeNotSet {
+                found: key.to_string(),
+                expected:  attributes_for_faceting
+                    .iter()
+                    .filter_map(|&id| schema.name(id))
+                    .map(str::to_string)
+                    .collect::<Vec<_>>() })
+        }
+        let value = split
+            .next()
+            .ok_or_else(|| FacetError::InvalidFormat(s.to_string()))?;
+        // unquoting the string if need be:
+        let mut indices = value.char_indices();
+        let value =  match (indices.next(), indices.last()) {
+            (Some((s, '\'')), Some((e, '\''))) |
+            (Some((s, '\"')), Some((e, '\"'))) => value[s + 1..e].to_string(),
+            _ => value.to_string(),
+        };
```

Comment:
Like I said before we should use a better parser like pest, we can't handle syntx errors like `["colors:'blue\""]`.
Also we do not support quoting the field name.